pkg/migrate/azure/privatedns.go

```diff
@@ -0,0 +1,744 @@
+	if err != nil {
+		return err
+	}
+
+	location := "global"
+	virtualNetworkID := *virtualNetwork.ID
+	virtualNetworkLinkName := fmt.Sprintf("%s-network-link", strings.Replace(client.vnetResourceGroup, "-rg", "", 1))
+
+	// pass a flag for this?
+	registrationEnabled := false
+
+	virtualNetworkLink := azprivatedns.VirtualNetworkLink{
+		Location: &location,
+		VirtualNetworkLinkProperties: &azprivatedns.VirtualNetworkLinkProperties{
+			VirtualNetwork: &azprivatedns.SubResource{
+				ID: &virtualNetworkID,
+			},
+			RegistrationEnabled: &registrationEnabled,
+		},
+	}
+
+	// Create the virtual network link to DNS
+	logrus.Printf("link: %s ... ", virtualNetworkLinkName)
+	linkFuture, err := client.virtualNetworkLinksClient.CreateOrUpdate(ctx, client.resourceGroup, *privateZone.Name, virtualNetworkLinkName, virtualNetworkLink, "", "")
+	if err != nil {
+		return err
+	}
+
+	// Wait for the link creation to complete
+	if err = linkFuture.WaitForCompletionRef(ctx, client.virtualNetworkLinksClient.Client); err != nil {
+		return err
+	}
+
+	// Read back the newly created link to verify creation
+	_, err = client.virtualNetworkLinksClient.Get(ctx, client.resourceGroup, *privateZone.Name, virtualNetworkLinkName)
+	if err != nil {
+		return err
+	}
+	logrus.Printf("ok.\n")
+
+	return nil
+}
+
+// Migrate does a migration from a legacy zone to a private zone
+func Migrate(oldResourceGroup string, newResourceGroup string, migrateZone string, virtualNetwork string, vnetResourceGroup string, link bool) error {
+	session, err := azconfig.GetSession()
+	if err != nil {
+		return err
+	}
+
+	if newResourceGroup == "" {
+		newResourceGroup = oldResourceGroup
+	}
+
+	legacyDNSClient := newLegacyDNSClient(session, oldResourceGroup)
+	privateDNSClient := newPrivateDNSClient(session, newResourceGroup, virtualNetwork, vnetResourceGroup)
+
+	// serialize and deserialize to make deep copies
+	szone, err := legacyDNSClient.GetZoneSerialized(migrateZone)
+	if err != nil {
+		return err
+	}
+
+	legacyDNSZoneInfo, err := deserializeLegacyZone(szone)
+	if err != nil {
+		return err
+	}
+
+	// create new private zone
+	err = privateDNSClient.MigrateLegacyZone(legacyDNSZoneInfo, link)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Eligible shows legacy zones that are eligible for migrating to private zones
+func Eligible() error {
+	session, err := azconfig.GetSession()
+	if err != nil {
+		return err
+	}
+
+	legacyDNSClient := newLegacyDNSClient(session, "")
+
+	szones, err := legacyDNSClient.GetZonesSerialized()
+	if err != nil {
+		return err
+	}
+
+	for _, szone := range szones {
+		legacyDNSZoneInfo, err := deserializeLegacyZone(szone)
+		if err != nil {
+			return err
+		}
+
+		eligibleZone := legacyDNSZoneInfo.zone
+		resourceGroup := idToResourceGroup(*eligibleZone.ID)
+
+		logrus.Printf("legacy zone=%s resourceGroup=%s\n", *eligibleZone.Name, resourceGroup)
```

Comment:
use logrus's fiedLogger to print these fields?

also why are you using Printf instead of `Debug(f)` or `Info(f)` ?

Reply:
No particular reason other than my personal preference. Printf() is same as Info(). I have updated it to use Info() to be consistent with the rest of the installer code, however.